src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,1631 @@
+    Keylet const& bridgeK,
+    STXChainBridge::ChainType const srcChain,
+    std::unordered_map<AccountID, std::uint32_t> const& signersList,
+    std::uint32_t quorum)
+{
+    if (attBegin == attEnd)
+        return tesSUCCESS;
+
+    PaymentSandbox psb(&ctx_.view());
+
+    auto const sleDoor = psb.peek(doorK);
+    if (!sleDoor)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const sleB = psb.peek(bridgeK);
+    if (!sleB)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    std::int64_t const claimCount = (*sleB)[sfXChainAccountClaimCount];
+
+    if (attBegin->createCount <= claimCount)
+    {
+        return tecXCHAIN_ACCOUNT_CREATE_PAST;
+    }
+    if (attBegin->createCount >= claimCount + 128)
+    {
+        // Limit the number of claims on the account
+        return tecXCHAIN_ACCOUNT_CREATE_TOO_MANY;
+    }
+
+    STXChainBridge::ChainType const dstChain =
+        STXChainBridge::otherChain(srcChain);
+
+    STXChainBridge::ChainType const attDstChain =
+        STXChainBridge::dstChain(attBegin->wasLockingChainSend);
+
+    if (attDstChain != dstChain)
+    {
+        return tecXCHAIN_WRONG_CHAIN;
+    }
+
+    auto const claimKeylet =
+        keylet::xChainCreateAccountClaimID(bridgeSpec, attBegin->createCount);
+
+    // sleCID may be null. If it's null it isn't created until the end of this
+    // function (if needed)
+    auto const sleCID = psb.peek(claimKeylet);
+    bool createCID = false;
+    if (!sleCID)
+    {
+        createCID = true;
+
+        // Check reserve
+        auto const balance = (*sleDoor)[sfBalance];
+        auto const reserve =
+            psb.fees().accountReserve((*sleDoor)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    std::vector<AttestationBatch::AttestationCreateAccount> atts;
+    atts.reserve(std::distance(attBegin, attEnd));
+    for (auto att = attBegin; att != attEnd; ++att)
+    {
+        if (!signersList.count(calcAccountID(att->publicKey)))
+            continue;
+        atts.push_back(*att);
+    }
+    if (atts.empty())
+    {
+        return tecXCHAIN_PROOF_UNKNOWN_KEY;
+    }
+
+    XChainCreateAccountAttestations curAtts = [&] {
+        if (sleCID)
+            return XChainCreateAccountAttestations{
+                sleCID->getFieldArray(sfXChainCreateAccountAttestations)};
+        return XChainCreateAccountAttestations{};
+    }();
+
+    auto const rewardAccounts = curAtts.onNewAttestations(
+        &atts[0], &atts[0] + atts.size(), quorum, signersList);
+
+    if (!createCID)
+    {
+        // Modify the object before it's potentially deleted, so the meta data
+        // will include the new attestations
+        if (!sleCID)
+            return tecINTERNAL;  // LCOV_EXCL_LINE
+        sleCID->setFieldArray(
+            sfXChainCreateAccountAttestations, curAtts.toSTArray());
+        psb.update(sleCID);
+    }
+
+    // Account create transactions must happen in order
```

Comment:
Over the weekend I had an idea on how to loosen the strict ordering restriction for the AccountCreate transactions.
Currently we store the claimCount in the Bridge ledger object and we allow only the next claimIdx. This is equivalent to the following C++ behavior:

```
// We process only AccountCreate claims in strict sequence. If one claim
// is not attested to for some reason, no further claim can be processed.
struct A
{
    uint32_t claimCount;

    bool
    canProcessClaim(uint32_t claimIdx) const
    {
        return claimIdx == claimCount + 1;
    }

    void
    registerClainProcessed(uint32_t claimIdx)
    {
        assert(canProcessClaim(claimIdx));
        ++claimCount;
    }
};
```

I think we could use an array of bools to allow processing accountCreate up to N ahead. Here is my C++ equivalent code:

```
// We allow to process AccountCreate claims within a window of arbitrary size N.
// This should be friendlier to out-of-order attestations or missing
// attestations.
struct B
{
    static constexpr uint32_t N = 20;

    uint32_t claimCount;
    std::array<bool, N> claimedAhead;  // initialized with false

    bool
    canProcessClaim(uint32_t claimIdx) const
    {
        return claimIdx > claimCount &&
            (claimIdx - claimCount < N && !claimedAhead[claimIdx - claimCount];
    }

    void
    registerClainProcessed(uint32_t claimIdx)
    {
        assert(canProcessClaim(claimIdx));
        if (claimIdx == claimCount + 1)
        {
            for (uint32_t i = 0; i < N - 1; ++i)
                claimedAhead[i] = claimedAhead[i + 1];
            claimedAhead[N - 1] = false;
            ++claimCount;
        }
        else
        {
            claimedAhead[claimIdx - claimCount] = true;
        }
    }
};
```

Reply:
We can also use this to prevent any locking of the accountCreate flow altogether, by failing the blocking AccountCreate and incrementing the `claimCount` if `claimIdx == claimCount + N` and the `claimedAhead` array is mostly true (maybe 50% threshold).

Reply:
I don't think we want to do this. If an account create get a quorum out of sequence, that's a bug in the witness servers. There will always be a limit to the number of transaction that could be unclogged. I prefer to keep the rippled implementation simple, possibly create a script to unclog the transaction if this happens, but I don't think the logic belongs on ledger.